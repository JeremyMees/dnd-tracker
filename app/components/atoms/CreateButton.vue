<script setup lang="ts">
defineEmits<{
  create: []
  hitLimit: []
}>()

defineProps<{
  allowCreate: boolean
  disabled?: boolean
  loading?: boolean
}>()
</script>

<template>
  <ClientOnly>
    <UiSkeleton
      v-if="loading"
      data-test-skeleton
      class="w-[120px] h-[47px]"
    />
    <button
      v-else
      class="btn-primary w-[120px]"
      :aria-label="$t('actions.create')"
      :disabled="disabled"
      @click="allowCreate ? $emit('create') : $emit('hitLimit')"
    >
      {{ $t('actions.create') }}
    </button>

    <template #fallback>
      <UiSkeleton class="w-[120px] h-[47px]" />
    </template>
  </ClientOnly>
</template>
