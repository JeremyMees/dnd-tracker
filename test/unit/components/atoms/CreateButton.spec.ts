import { mountSuspended } from '@nuxt/test-utils/runtime'
import { describe, expect, it } from 'vitest'
import { nextTick } from 'vue'
import CreateButton from '~/components/atoms/CreateButton.vue'

interface Props {
  allowCreate: boolean
  disabled?: boolean
  loading?: boolean
}

const props: Props = {
  allowCreate: true,
}

describe('CreateButton', async () => {
  it('Should match snapshot', async () => {
    const component = await mountSuspended(CreateButton, { props })

    expect(component.html()).toMatchSnapshot()
  })

  it('Should render with default props correctly', async () => {
    const component = await mountSuspended(CreateButton, { props })
    const button = component.find('button')

    expect(button.exists()).toBeTruthy()
    expect(button.text()).toBe('actions.create')
    expect(button.attributes('aria-label')).toBe('actions.create')
  })

  it('Should render disabled when disabled prop is true', async () => {
    const component = await mountSuspended(CreateButton, { props: { ...props, disabled: true } })
    const button = component.find('button')

    await button.trigger('click')
    await nextTick()

    expect(component.emitted('create')).toBeFalsy()
    expect(button.attributes('disabled')).toBeDefined()
  })

  it('Should render a loading state', async () => {
    const component = await mountSuspended(CreateButton, { props: { ...props, loading: true } })

    expect(component.find('[data-test-skeleton]').exists()).toBeTruthy()
    expect(component.find('button').exists()).toBeFalsy()
  })

  it('Should emit create event when clicked and allowed', async () => {
    const component = await mountSuspended(CreateButton, { props })

    await component.find('button').trigger('click')

    const emitted = component.emitted('create')
    expect(emitted).toBeTruthy()
    expect(emitted).toHaveLength(1)
  })

  it('Should emit hitLimit event when clicked but not allowed', async () => {
    const component = await mountSuspended(CreateButton, { props: { allowCreate: false } })

    await component.find('button').trigger('click')
    await nextTick()

    const emitted = component.emitted('hitLimit')
    expect(emitted).toBeTruthy()
    expect(emitted).toHaveLength(1)
  })
})
