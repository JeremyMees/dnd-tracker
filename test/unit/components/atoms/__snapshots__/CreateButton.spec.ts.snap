// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CreateButton > Should match snapshot 1`] = `"<button class="btn-primary w-[120px]" aria-label="actions.create">actions.create</button>"`;
